test/rand_test.c

```diff
@@ -76,11 +76,77 @@ static int test_rand_uniform(void)
     return res;
 }
 
+typedef struct r_test_ctx {
+    const OSSL_CORE_HANDLE *handle;
+} R_TEST_CTX;
+
+static void r_teardown(void *provctx)
+{
+    R_TEST_CTX *ctx = (R_TEST_CTX *)provctx;
+
+    free(ctx);
+}
+
+static int r_random(ossl_unused void *vprov, ossl_unused int which,
+                    void *buf, size_t n, ossl_unused unsigned int strength)
+{
+    while (n-- > 0)
+        ((unsigned char *)buf)[n] = 0xff & n;
+    return 1;
+}
+
+static const OSSL_DISPATCH r_test_table[] = {
+    { OSSL_FUNC_PROVIDER_RANDOM, (void (*)(void))r_random },
+    { OSSL_FUNC_PROVIDER_TEARDOWN, (void (*)(void))r_teardown },
+    OSSL_DISPATCH_END
+};
+
+static int r_init(const OSSL_CORE_HANDLE *handle,
+                  ossl_unused const OSSL_DISPATCH *oin,
+                  const OSSL_DISPATCH **out,
+                  void **provctx)
+{
+    R_TEST_CTX *ctx;
+
+    ctx = malloc(sizeof(*ctx));
+    if (ctx == NULL)
+        return 0;
+    ctx->handle = handle;
+
+    *provctx = (void *)ctx;
+    *out = r_test_table;
+    return 1;
+}
+
+static int test_rand_random_proivider(void)
+{
+    OSSL_LIB_CTX *ctx = NULL;
+    OSSL_PROVIDER *prov = NULL;
+    int res = 0;
+    static const unsigned char data[] = { 0, 1, 2, 3, 4, 5, 6 };
+    unsigned char buf[sizeof(data)];
+
+    if (!test_get_libctx(&ctx, NULL, NULL, NULL, NULL)
+            || !TEST_true(OSSL_PROVIDER_add_builtin(ctx, "r_prov", &r_init))
+            || !TEST_ptr(prov = OSSL_PROVIDER_try_load(ctx, "r_prov", 1))
+            || !TEST_true(RAND_set1_random_provider(ctx, prov))
+            || !RAND_bytes_ex(ctx, buf, sizeof(buf), 256)
+            || !TEST_mem_eq(buf, sizeof(buf), data, sizeof(data)))
+        goto err;
+
+    res = 1;
+ err:
+    OSSL_PROVIDER_unload(prov);
+    OSSL_LIB_CTX_free(ctx);
+    return res;
+}
+
 int setup_tests(void)
 {
     if (!TEST_true(RAND_set_DRBG_type(NULL, "TEST-RAND", NULL, NULL, NULL)))
         return 0;
     ADD_TEST(test_rand);
     ADD_TEST(test_rand_uniform);
+    ADD_TEST(test_rand_random_proivider);
```

Comment:
Should there be a test somewhere to reset the random provider back to default?

Should there be a test to confirm interactions with the fips provider are as we expect?

Reply:
I'm not sure how to test the latter.  All we've got to work with is the output random bytes which we're going to get both with and without the FIPS provider.  I've verified that it is working but there is no obvious way to test such directly.  The test here verifies that the framework around this is functional, I'm not sure how to do better.

A test to verify that this can be disabled is worthwhile and possible.